File: sol/solution/0001-0100/0037/Solution.rb
def solve_sudoku(board)
    rows = Array.new(9, 0)
    cols = Array.new(9, 0)
    boxes = Array.new(9, 0)
    empties = []
    9.times do |r|
        9.times do |c|
            ch = board[r][c]
            if ch == '.'
                empties << [r, c]
            else
                bit = 1 << (ch.ord - '1'.ord)
                rows[r] |= bit
                cols[c] |= bit
                boxes[(r/3)*3 + c/3] |= bit
            end
        end
    end
    backtrack = lambda do
        target_idx = -1
        candidates_mask = 0
        min_count = 10
        empties.each_with_index do |(r, c), idx|
            next unless board[r][c] == '.'
            used = rows[r] | cols[c] | boxes[(r/3)*3 + c/3]
            mask = (~used) & 0x1FF
            return false if mask == 0
            count = 0
            m = mask
            while m > 0
                m &= m - 1
                count += 1
            end
            if count < min_count
                min_count = count
                target_idx = idx
                candidates_mask = mask
                break if count == 1
            end
        end
        return true if target_idx == -1
        r, c = empties[target_idx]
        box_idx = (r/3)*3 + c/3
        mask = candidates_mask
        1.upto(9) do |d|
            bit = 1 << (d - 1)
            next if (mask & bit) == 0
            board[r][c] = d.to_s
            rows[r] |= bit
            cols[c] |= bit
            boxes[box_idx] |= bit
            return true if backtrack.call
            board[r][c] = '.'
            rows[r] ^= bit
            cols[c] ^= bit
            boxes[box_idx] ^= bit
        end
        false
    end
    backtrack.call
end

File: sol/solution/0001-0100/0037/Solution.php
class Solution {
    function solveSudoku(&$board) {
        $rows = array_fill(0, 9, 0);
        $cols = array_fill(0, 9, 0);
        $boxes = array_fill(0, 9, 0);
        $empties = [];
        for ($r = 0; $r < 9; $r++) {
            for ($c = 0; $c < 9; $c++) {
                if ($board[$r][$c] === '.') {
                    $empties[] = [$r, $c];
                } else {
                    $d = intval($board[$r][$c]) - 1;
                    $mask = 1 << $d;
                    $rows[$r] |= $mask;
                    $cols[$c] |= $mask;
                    $boxes[$this->boxIndex($r,$c)] |= $mask;
                }
            }
        }
        $this->dfs($board, $empties, 0, $rows, $cols, $boxes);
    }
    private function boxIndex($r,$c) {
        return intdiv($r,3)*3 + intdiv($c,3);
    }
    private function dfs(&$board, &$empties, $pos, &$rows, &$cols, &$boxes) {
        if ($pos === count($empties)) return true;
        $best = $pos;
        $bestCount = 10;
        for ($k = $pos; $k < count($empties); $k++) {
            [$r,$c] = $empties[$k];
            $b = $this->boxIndex($r,$c);
            $used = $rows[$r] | $cols[$c] | $boxes[$b];
            $cand = (~$used) & 0x1FF;
            $count = $this->bitCount($cand);
            if ($count < $bestCount) {
                $bestCount = $count;
                $best = $k;
                if ($count === 1) break;
            }
        }
        if ($best !== $pos) {
            $tmp = $empties[$pos];
            $empties[$pos] = $empties[$best];
            $empties[$best] = $tmp;
        }
        [$r,$c] = $empties[$pos];
        $b = $this->boxIndex($r,$c);
        $used = $rows[$r] | $cols[$c] | $boxes[$b];
        $cand = (~$used) & 0x1FF;
        if ($cand === 0) return false;
        for ($d = 0; $d < 9; $d++) {
            $bit = 1 << $d;
            if (!($cand & $bit)) continue;
            $rows[$r] |= $bit;
            $cols[$c] |= $bit;
            $boxes[$b] |= $bit;
            $board[$r][$c] = strval($d+1);
            if ($this->dfs($board, $empties, $pos+1, $rows, $cols, $boxes)) return true;
            $rows[$r] ^= $bit;
            $cols[$c] ^= $bit;
            $boxes[$b] ^= $bit;
            $board[$r][$c] = '.';
        }
        return false;
    }
    private function bitCount($x) {
        $c = 0;
        while ($x) { $x &= $x - 1; $c++; }
        return $c;
    }
}
